compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs

```diff
@@ -706,4 +700,33 @@ fn provide_cstore_hooks(providers: &mut Providers) {
             cdata.imported_source_file(file_index as u32, tcx.sess);
         }
     };
+    providers.hooks.allocator_kind = |tcx| CStore::from_tcx(tcx).allocator_kind();
+    providers.hooks.alloc_error_handler_kind =
+        |tcx| CStore::from_tcx(tcx).alloc_error_handler_kind();
+    providers.hooks.has_global_allocator = |tcx, krate: CrateNum| {
+        if krate == LOCAL_CRATE {
+            CStore::from_tcx(tcx).has_global_allocator()
+        } else {
+            let cdata =
+                rustc_data_structures::sync::FreezeReadGuard::map(CStore::from_tcx(tcx), |c| {
+                    c.get_crate_data(krate).cdata
+                });
+            let cdata =
+                crate::creader::CrateMetadataRef { cdata: &cdata, cstore: &CStore::from_tcx(tcx) };
+            cdata.root.has_global_allocator
+        }
+    };
+    providers.hooks.has_alloc_error_handler = |tcx, krate: CrateNum| {
+        if krate == LOCAL_CRATE {
+            CStore::from_tcx(tcx).has_alloc_error_handler()
+        } else {
+            let cdata =
+                rustc_data_structures::sync::FreezeReadGuard::map(CStore::from_tcx(tcx), |c| {
+                    c.get_crate_data(krate).cdata
```

Comment:
```suggestion
                    c.get_crate_data(krate).cdata.root.has_alloc_error_handler
```
No need to create `CrateMetadataRef` below.